docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1680 @@
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            module M where
+
+            data A = A
+              with
+
+     - .. code-block:: daml
+
+            module M where
+
+            data A = A 
+              with 
+                x : Optional (Int -> Int)
+
+Data Types - Records
+~~~~~~~~~~~~~~~~~~~~
+
+An upgrading record may add new optional fields at the end of the field
+sequence of the upgraded record. The types of the fields that the
+upgrading record has in common with the upgraded record must be pairwise
+valid upgrades of the original types.
+
+Deleting a field leads to a validation error.
+
+Adding a field in the middle of the field sequence leads to a validation
+error.
+
+As a special case of the two points above, renaming a field leads to a
+validation error.
+
+Adding a non-optional field at the end of the field sequence leads to a
+validation error.
+
+.. _examples-7:
+
+**Examples**
+
+Below, the record on the right is a valid upgrade of the module on the
+left. It adds an optional field ``x2`` at the end of the field sequence.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+          data T = T with
+            x1 : Int
+
+     - .. code-block:: daml
+
+          data T = T with
+           x1 : Int
+           x2 : Optional Text
+
+Below, the record on the right is **not** a valid upgrade of the record
+on the left because it adds a new field ``x2`` before ``x1`` instead of adding
+it at the end of the field sequence.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+          data T = T with
+            x1 : Int
+
+     - .. code-block:: daml
+
+          data T = T with
+            x2 : Optional Text
+            x1 : Int
+  
+Below, the record on the right is **not** a valid upgrade of the record
+on the left because it drops field ``x2``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+           data T = T with
+             x1 : Int
+             x2 : Text
+     
+     - .. code-block:: daml
+
+           data T = T with
+             x1 : Int
+
+Below, the record on the right is **not** a valid upgrade of the record
+on the left because it changes the type of ``x1`` from ``Int`` to ``Text``. 
+``Text`` is not a valid upgrade of ``Int``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+           data T = T with
+             x1 : Int
+
+     - .. code-block:: daml
+
+           data T = T with
+             x1 : Text
+
+Data Types - Variants
+~~~~~~~~~~~~~~~~~~~~~
+
+An upgrading variant may add new constructors at the end of the
+constructor sequence of the upgraded variant. The argument types  
+of the constructors that the upgrading variant has in common with the  
+upgraded variant must be pairwise valid upgrades of the original types.  
```

Comment:
Not clarified here: 
1. Can I add another field to a constructor at the end of its argument sequence? 
2. What happens if I want to add a field to a nullary constructor?
3. What considerations are there wrt avoiding to convert enums into variants?

Reply:
Addressed all three points and added examples.